File: Legacy/Source/modForum.vb
Option Strict Off
Option Explicit On
Module modForum
    Public Const MAX_MENSAJES_FORO As Byte = 30
    Public Const MAX_ANUNCIOS_FORO As Byte = 5

    Public Const FORO_REAL_ID As String = "REAL"
    Public Const FORO_CAOS_ID As String = "CAOS"

    Public Structure tPost
        Dim sTitulo As String
        Dim sPost As String
        Dim Autor As String
    End Structure

    Public Structure tForo
        <VBFixedArray(MAX_MENSAJES_FORO)> Dim vsPost() As tPost
        <VBFixedArray(MAX_ANUNCIOS_FORO)> Dim vsAnuncio() As tPost
        Dim CantPosts As Byte
        Dim CantAnuncios As Byte
        Dim ID As String

        'UPGRADE_TODO: Se debe llamar a "Initialize" para inicializar instancias de esta estructura. Haga clic aquí para obtener más información: 'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="B4BFF9E0-8631-45CF-910E-62AB3970F27B"'
        Public Sub Initialize()
            'UPGRADE_WARNING: El límite inferior de la matriz vsPost ha cambiado de 1 a 0. Haga clic aquí para obtener más información: 'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="0F1C9BE1-AF9D-476E-83B1-17D43BECFF20"'
            ReDim vsPost(MAX_MENSAJES_FORO)
            'UPGRADE_WARNING: El límite inferior de la matriz vsAnuncio ha cambiado de 1 a 0. Haga clic aquí para obtener más información: 'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="0F1C9BE1-AF9D-476E-83B1-17D43BECFF20"'
            ReDim vsAnuncio(MAX_ANUNCIOS_FORO)
        End Sub
    End Structure

    Private NumForos As Short
    Private Foros() As tForo


    Public Sub AddForum(sForoID As String)
        '***************************************************
        'Author: ZaMa
        'Last Modification: 22/02/2010
        'Adds a forum to the list and fills it.
        '***************************************************
        Dim ForumPath As String
        Dim PostPath As String
        Dim PostIndex As Short
        Dim FileIndex As Short

        NumForos = NumForos + 1
        'UPGRADE_WARNING: Es posible que la matriz Foros necesite tener elementos individuales inicializados. Haga clic aquí para obtener más información: 'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="B97B714D-9338-48AC-B03F-345B617E2B02"'
        'UPGRADE_WARNING: El límite inferior de la matriz Foros ha cambiado de 1 a 0. Haga clic aquí para obtener más información: 'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="0F1C9BE1-AF9D-476E-83B1-17D43BECFF20"'
        ReDim Preserve Foros(NumForos)
        InitializeStruct(Foros)

        ForumPath = AppDomain.CurrentDomain.BaseDirectory & "foros/" & sForoID & ".for"

        With Foros(NumForos)

            .ID = sForoID

            If FileExist(ForumPath) Then
                .CantPosts = Val(GetVar(ForumPath, "INFO", "CantMSG"))
                .CantAnuncios = Val(GetVar(ForumPath, "INFO", "CantAnuncios"))

                ' Cargo posts
                For PostIndex = 1 To .CantPosts
                    FileIndex = FreeFile
                    PostPath = AppDomain.CurrentDomain.BaseDirectory & "foros/" & sForoID & PostIndex & ".for"

                    FileOpen(FileIndex, PostPath, OpenMode.Input, , OpenShare.Shared)

                    ' Titulo
                    Input(FileIndex, .vsPost(PostIndex).sTitulo)
                    ' Autor
                    Input(FileIndex, .vsPost(PostIndex).Autor)
                    ' Mensaje
                    Input(FileIndex, .vsPost(PostIndex).sPost)

                    FileClose(FileIndex)
                Next PostIndex

                ' Cargo anuncios
                For PostIndex = 1 To .CantAnuncios
                    FileIndex = FreeFile
                    PostPath = AppDomain.CurrentDomain.BaseDirectory & "foros/" & sForoID & PostIndex & "a.for"

                    FileOpen(FileIndex, PostPath, OpenMode.Input, , OpenShare.Shared)

                    ' Titulo
                    Input(FileIndex, .vsAnuncio(PostIndex).sTitulo)
                    ' Autor
                    Input(FileIndex, .vsAnuncio(PostIndex).Autor)
                    ' Mensaje
                    Input(FileIndex, .vsAnuncio(PostIndex).sPost)

                    FileClose(FileIndex)
                Next PostIndex
            End If

        End With
    End Sub

    Public Function GetForumIndex(ByRef sForoID As String) As Short
        '***************************************************
        'Author: ZaMa
        'Last Modification: 22/02/2010
        'Returns the forum index.
        '***************************************************

        Dim ForumIndex As Short

        For ForumIndex = 1 To NumForos
            If Foros(ForumIndex).ID = sForoID Then
                GetForumIndex = ForumIndex
                Exit Function
            End If
        Next ForumIndex
    End Function

    Public Sub AddPost(ForumIndex As Short, ByRef Post As String, ByRef Autor As String, ByRef Titulo As String,
                       bAnuncio As Boolean)
        '***************************************************
        'Author: ZaMa
        'Last Modification: 22/02/2010
        'Saves a new post into the forum.
        '***************************************************

        With Foros(ForumIndex)

            If bAnuncio Then
                If .CantAnuncios < MAX_ANUNCIOS_FORO Then .CantAnuncios = .CantAnuncios + 1

                Call MoveArray(ForumIndex, bAnuncio)

                ' Agrego el anuncio
                With .vsAnuncio(1)
                    .sTitulo = Titulo
                    .Autor = Autor
                    .sPost = Post
                End With

            Else
                If .CantPosts < MAX_MENSAJES_FORO Then .CantPosts = .CantPosts + 1

                Call MoveArray(ForumIndex, bAnuncio)

                ' Agrego el post
                With .vsPost(1)
                    .sTitulo = Titulo
                    .Autor = Autor
                    .sPost = Post
                End With

            End If
        End With
    End Sub

    Public Sub SaveForums()
        '***************************************************
        'Author: ZaMa
        'Last Modification: 22/02/2010
        'Saves all forums into disk.
        '***************************************************
        Dim ForumIndex As Short

        For ForumIndex = 1 To NumForos
            Call SaveForum(ForumIndex)
        Next ForumIndex
    End Sub


    Private Sub SaveForum(ForumIndex As Short)
        '***************************************************
        'Author: ZaMa
        'Last Modification: 22/02/2010
        'Saves a forum into disk.
        '***************************************************

        Dim PostIndex As Short
        Dim FileIndex As Short
        Dim PostPath As String

        Call CleanForum(ForumIndex)

        With Foros(ForumIndex)

            ' Guardo info del foro
            Call _
                WriteVar(AppDomain.CurrentDomain.BaseDirectory & "Foros/" & .ID & ".for", "INFO", "CantMSG",
                         CStr(.CantPosts))
            Call _
                WriteVar(AppDomain.CurrentDomain.BaseDirectory & "Foros/" & .ID & ".for", "INFO", "CantAnuncios",
                         CStr(.CantAnuncios))

            ' Guardo posts
            For PostIndex = 1 To .CantPosts

                PostPath = AppDomain.CurrentDomain.BaseDirectory & "Foros/" & .ID & PostIndex & ".for"
                FileIndex = FreeFile
                FileOpen(FileIndex, PostPath, OpenMode.Output)

                With .vsPost(PostIndex)
                    PrintLine(FileIndex, .sTitulo)
                    PrintLine(FileIndex, .Autor)
                    PrintLine(FileIndex, .sPost)
                End With

                FileClose(FileIndex)

            Next PostIndex

            ' Guardo Anuncios
            For PostIndex = 1 To .CantAnuncios

                PostPath = AppDomain.CurrentDomain.BaseDirectory & "Foros/" & .ID & PostIndex & "a.for"
                FileIndex = FreeFile
                FileOpen(FileIndex, PostPath, OpenMode.Output)

                With .vsAnuncio(PostIndex)
                    PrintLine(FileIndex, .sTitulo)
                    PrintLine(FileIndex, .Autor)
                    PrintLine(FileIndex, .sPost)
                End With

                FileClose(FileIndex)

            Next PostIndex

        End With
    End Sub

    Public Sub CleanForum(ForumIndex As Short)
        '***************************************************
        'Author: ZaMa
        'Last Modification: 22/02/2010
        'Cleans a forum from disk.
        '***************************************************
        Dim PostIndex As Short
        Dim NumPost As Short
        Dim ForumPath As String

        With Foros(ForumIndex)

            ' Elimino todo
            ForumPath = AppDomain.CurrentDomain.BaseDirectory & "Foros/" & .ID & ".for"
            If FileExist(ForumPath) Then

                NumPost = Val(GetVar(ForumPath, "INFO", "CantMSG"))

                ' Elimino los post viejos
                For PostIndex = 1 To NumPost
                    Kill(AppDomain.CurrentDomain.BaseDirectory & "Foros/" & .ID & PostIndex & ".for")
                Next PostIndex


                NumPost = Val(GetVar(ForumPath, "INFO", "CantAnuncios"))

                ' Elimino los post viejos
                For PostIndex = 1 To NumPost
                    Kill(AppDomain.CurrentDomain.BaseDirectory & "Foros/" & .ID & PostIndex & "a.for")
                Next PostIndex


                ' Elimino el foro
                Kill(AppDomain.CurrentDomain.BaseDirectory & "Foros/" & .ID & ".for")

            End If
        End With
    End Sub

    Public Function SendPosts(UserIndex As Short, ByRef ForoID As String) As Boolean
        '***************************************************
        'Author: ZaMa
        'Last Modification: 22/02/2010
        'Sends all the posts of a required forum
        '***************************************************

        Dim ForumIndex As Short
        Dim PostIndex As Short
        Dim bEsGm As Boolean

        ForumIndex = GetForumIndex(ForoID)

        If ForumIndex > 0 Then

            With Foros(ForumIndex)

                ' Send General posts
                For PostIndex = 1 To .CantPosts
                    With .vsPost(PostIndex)
                        Call _
                            WriteAddForumMsg(UserIndex, eForumMsgType.ieGeneral, .sTitulo, .Autor, .sPost)
                    End With
                Next PostIndex

                ' Send Sticky posts
                For PostIndex = 1 To .CantAnuncios
                    With .vsAnuncio(PostIndex)
                        Call _
                            WriteAddForumMsg(UserIndex, eForumMsgType.ieGENERAL_STICKY, .sTitulo, .Autor,
                                             .sPost)
                    End With
                Next PostIndex

            End With

            bEsGm = EsGM(UserIndex)

            ' Caos?
            If esCaos(UserIndex) Or bEsGm Then

                ForumIndex = GetForumIndex(FORO_CAOS_ID)

                With Foros(ForumIndex)

                    ' Send General Caos posts
                    For PostIndex = 1 To .CantPosts

                        With .vsPost(PostIndex)
                            Call _
                                WriteAddForumMsg(UserIndex, eForumMsgType.ieCAOS, .sTitulo, .Autor, .sPost)
                        End With

                    Next PostIndex

                    ' Send Sticky posts
                    For PostIndex = 1 To .CantAnuncios
                        With .vsAnuncio(PostIndex)
                            Call _
                                WriteAddForumMsg(UserIndex, eForumMsgType.ieCAOS_STICKY, .sTitulo, .Autor,
                                                 .sPost)
                        End With
                    Next PostIndex

                End With
            End If

            ' Caos?
            If esArmada(UserIndex) Or bEsGm Then

                ForumIndex = GetForumIndex(FORO_REAL_ID)

                With Foros(ForumIndex)

                    ' Send General Real posts
                    For PostIndex = 1 To .CantPosts

                        With .vsPost(PostIndex)
                            Call _
                                WriteAddForumMsg(UserIndex, eForumMsgType.ieREAL, .sTitulo, .Autor, .sPost)
                        End With

                    Next PostIndex

                    ' Send Sticky posts
                    For PostIndex = 1 To .CantAnuncios
                        With .vsAnuncio(PostIndex)
                            Call _
                                WriteAddForumMsg(UserIndex, eForumMsgType.ieREAL_STICKY, .sTitulo, .Autor,
                                                 .sPost)
                        End With
                    Next PostIndex

                End With
            End If

            SendPosts = True
        End If
    End Function

    Public Function EsAnuncio(ForumType As Byte) As Boolean
        '***************************************************
        'Author: ZaMa
        'Last Modification: 22/02/2010
        'Returns true if the post is sticky.
        '***************************************************
        Select Case ForumType
            Case eForumMsgType.ieCAOS_STICKY
                EsAnuncio = True

            Case eForumMsgType.ieGENERAL_STICKY
                EsAnuncio = True

            Case eForumMsgType.ieREAL_STICKY
                EsAnuncio = True

        End Select
    End Function

    Public Function ForumAlignment(yForumType As Byte) As Byte
        '***************************************************
        'Author: ZaMa
        'Last Modification: 01/03/2010
        'Returns the forum alignment.
        '***************************************************
        Select Case yForumType
            Case eForumMsgType.ieCAOS, eForumMsgType.ieCAOS_STICKY
                ForumAlignment = eForumType.ieCAOS

            Case eForumMsgType.ieGeneral, eForumMsgType.ieGENERAL_STICKY
                ForumAlignment = eForumType.ieGeneral

            Case eForumMsgType.ieREAL, eForumMsgType.ieREAL_STICKY
                ForumAlignment = eForumType.ieREAL

        End Select
    End Function

    Public Sub ResetForums()
        '***************************************************
        'Author: ZaMa
        'Last Modification: 22/02/2010
        'Resets forum info
        '***************************************************
        'UPGRADE_WARNING: Es posible que la matriz Foros necesite tener elementos individuales inicializados. Haga clic aquí para obtener más información: 'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="B97B714D-9338-48AC-B03F-345B617E2B02"'
        'UPGRADE_WARNING: El límite inferior de la matriz Foros ha cambiado de 1 a 0. Haga clic aquí para obtener más información: 'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="0F1C9BE1-AF9D-476E-83B1-17D43BECFF20"'
        ReDim Foros(1)
        InitializeStruct(Foros)
        NumForos = 0
    End Sub

    Private Sub MoveArray(ForumIndex As Short, Sticky As Boolean)
        Dim i As Integer

        With Foros(ForumIndex)
            If Sticky Then
                For i = .CantAnuncios To 2 Step - 1
                    .vsAnuncio(i).sTitulo = .vsAnuncio(i - 1).sTitulo
                    .vsAnuncio(i).sPost = .vsAnuncio(i - 1).sPost
                    .vsAnuncio(i).Autor = .vsAnuncio(i - 1).Autor
                Next i
            Else
                For i = .CantPosts To 2 Step - 1
                    .vsPost(i).sTitulo = .vsPost(i - 1).sTitulo
                    .vsPost(i).sPost = .vsPost(i - 1).sPost
                    .vsPost(i).Autor = .vsPost(i - 1).Autor
                Next i
            End If
        End With
    End Sub
End Module

File: Legacy/Source/modForum.cs
using System;
using System.IO;
using Microsoft.VisualBasic;

namespace Legacy;

internal static class modForum
{
    public const byte MAX_MENSAJES_FORO = 30;
    public const byte MAX_ANUNCIOS_FORO = 5;

    public const string FORO_REAL_ID = "REAL";
    public const string FORO_CAOS_ID = "CAOS";

    private static short NumForos;
    private static tForo[] Foros;


    public static void AddForum(string sForoID)
    {
        // ***************************************************
        // Author: ZaMa
        // Last Modification: 22/02/2010
        // Adds a forum to the list and fills it.
        // ***************************************************
        string ForumPath;
        string PostPath;
        short PostIndex;
        short FileIndex;

        NumForos = Convert.ToInt16(NumForos + 1);
        // UPGRADE_WARNING: Es posible que la matriz Foros necesite tener elementos individuales inicializados. Haga clic aquí para obtener más información: 'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="B97B714D-9338-48AC-B03F-345B617E2B02"'
        // UPGRADE_WARNING: El límite inferior de la matriz Foros ha cambiado de 1 a 0. Haga clic aquí para obtener más información: 'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="0F1C9BE1-AF9D-476E-83B1-17D43BECFF20"'
        Array.Resize(ref Foros, NumForos + 1);
        ArrayInitializers.InitializeStruct(Foros);

        ForumPath = AppDomain.CurrentDomain.BaseDirectory + "foros/" + sForoID + ".for";

        {
            ref var withBlock = ref Foros[NumForos];

            withBlock.ID = sForoID;

            if (File.Exists(ForumPath))
            {
                var argEmptySpaces = 1024;
                withBlock.CantPosts =
                    Convert.ToByte(Migration.ParseVal(ES.GetVar(ForumPath, "INFO", "CantMSG", ref argEmptySpaces)));
                var argEmptySpaces1 = 1024;
                withBlock.CantAnuncios =
                    Convert.ToByte(
                        Migration.ParseVal(ES.GetVar(ForumPath, "INFO", "CantAnuncios", ref argEmptySpaces1)));

                // Cargo posts
                var loopTo = (short)withBlock.CantPosts;
                for (PostIndex = 1; PostIndex <= loopTo; PostIndex++)
                {
                    PostPath = AppDomain.CurrentDomain.BaseDirectory + "foros/" + sForoID + PostIndex + ".for";

                    using (var reader = new StreamReader(PostPath))
                    {
                        // Titulo
                        withBlock.vsPost[PostIndex].sTitulo = reader.ReadLine();
                        // Autor
                        withBlock.vsPost[PostIndex].Autor = reader.ReadLine();
                        // Mensaje
                        withBlock.vsPost[PostIndex].sPost = reader.ReadLine();
                    }
                }

                // Cargo anuncios
                var loopTo1 = (short)withBlock.CantAnuncios;
                for (PostIndex = 1; PostIndex <= loopTo1; PostIndex++)
                {
                    PostPath = AppDomain.CurrentDomain.BaseDirectory + "foros/" + sForoID + PostIndex + "a.for";

                    using (var reader = new StreamReader(PostPath))
                    {
                        // Titulo
                        withBlock.vsAnuncio[PostIndex].sTitulo = reader.ReadLine();
                        // Autor
                        withBlock.vsAnuncio[PostIndex].Autor = reader.ReadLine();
                        // Mensaje
                        withBlock.vsAnuncio[PostIndex].sPost = reader.ReadLine();
                    }
                }
            }
        }
    }

    public static short GetForumIndex(ref string sForoID)
    {
        short GetForumIndexRet = default;
        // ***************************************************
        // Author: ZaMa
        // Last Modification: 22/02/2010
        // Returns the forum index.
        // ***************************************************

        short ForumIndex;

        var loopTo = NumForos;
        for (ForumIndex = 1; ForumIndex <= loopTo; ForumIndex++)
            if ((Foros[ForumIndex].ID ?? "") == (sForoID ?? ""))
            {
                GetForumIndexRet = ForumIndex;
                return GetForumIndexRet;
            }

        return GetForumIndexRet;
    }

    public static void AddPost(short ForumIndex, ref string Post, ref string Autor, ref string Titulo, bool bAnuncio)
    {
        // ***************************************************
        // Author: ZaMa
        // Last Modification: 22/02/2010
        // Saves a new post into the forum.
        // ***************************************************

        {
            ref var withBlock = ref Foros[ForumIndex];

            if (bAnuncio)
            {
                if (withBlock.CantAnuncios < MAX_ANUNCIOS_FORO)
                    withBlock.CantAnuncios = Convert.ToByte(withBlock.CantAnuncios + 1);

                MoveArray(ForumIndex, bAnuncio);

                // Agrego el anuncio
                {
                    ref var withBlock1 = ref withBlock.vsAnuncio[1];
                    withBlock1.sTitulo = Titulo;
                    withBlock1.Autor = Autor;
                    withBlock1.sPost = Post;
                }
            }

            else
            {
                if (withBlock.CantPosts < MAX_MENSAJES_FORO)
                    withBlock.CantPosts = Convert.ToByte(withBlock.CantPosts + 1);

                MoveArray(ForumIndex, bAnuncio);

                // Agrego el post
                {
                    ref var withBlock2 = ref withBlock.vsPost[1];
                    withBlock2.sTitulo = Titulo;
                    withBlock2.Autor = Autor;
                    withBlock2.sPost = Post;
                }
            }
        }
    }

    public static void SaveForums()
    {
        // ***************************************************
        // Author: ZaMa
        // Last Modification: 22/02/2010
        // Saves all forums into disk.
        // ***************************************************
        short ForumIndex;

        var loopTo = NumForos;
        for (ForumIndex = 1; ForumIndex <= loopTo; ForumIndex++)
            SaveForum(ForumIndex);
    }


    private static void SaveForum(short ForumIndex)
    {
        // ***************************************************
        // Author: ZaMa
        // Last Modification: 22/02/2010
        // Saves a forum into disk.
        // ***************************************************

        short PostIndex;
        short FileIndex;
        string PostPath;

        CleanForum(ForumIndex);

        {
            ref var withBlock = ref Foros[ForumIndex];

            // Guardo info del foro
            ES.WriteVar(AppDomain.CurrentDomain.BaseDirectory + "Foros/" + withBlock.ID + ".for", "INFO", "CantMSG",
                withBlock.CantPosts.ToString());
            ES.WriteVar(AppDomain.CurrentDomain.BaseDirectory + "Foros/" + withBlock.ID + ".for", "INFO",
                "CantAnuncios", withBlock.CantAnuncios.ToString());

            // Guardo posts
            var loopTo = (short)withBlock.CantPosts;
            for (PostIndex = 1; PostIndex <= loopTo; PostIndex++)
            {
                PostPath = AppDomain.CurrentDomain.BaseDirectory + "Foros/" + withBlock.ID + PostIndex + ".for";
                using (var writer = new StreamWriter(PostPath))
                {
                    writer.WriteLine(withBlock.vsPost[PostIndex].sTitulo);
                    writer.WriteLine(withBlock.vsPost[PostIndex].Autor);
                    writer.WriteLine(withBlock.vsPost[PostIndex].sPost);
                }
            }

            // Guardo Anuncios
            var loopTo1 = (short)withBlock.CantAnuncios;
            for (PostIndex = 1; PostIndex <= loopTo1; PostIndex++)
            {
                PostPath = AppDomain.CurrentDomain.BaseDirectory + "Foros/" + withBlock.ID + PostIndex + "a.for";
                using (var writer = new StreamWriter(PostPath))
                {
                    writer.WriteLine(withBlock.vsAnuncio[PostIndex].sTitulo);
                    writer.WriteLine(withBlock.vsAnuncio[PostIndex].Autor);
                    writer.WriteLine(withBlock.vsAnuncio[PostIndex].sPost);
                }
            }
        }
    }

    public static void CleanForum(short ForumIndex)
    {
        // ***************************************************
        // Author: ZaMa
        // Last Modification: 22/02/2010
        // Cleans a forum from disk.
        // ***************************************************
        short PostIndex;
        short NumPost;
        string ForumPath;

        {
            ref var withBlock = ref Foros[ForumIndex];

            // Elimino todo
            ForumPath = AppDomain.CurrentDomain.BaseDirectory + "Foros/" + withBlock.ID + ".for";
            if (File.Exists(ForumPath))
            {
                var argEmptySpaces = 1024;
                NumPost = Convert.ToInt16(
                    Migration.ParseVal(ES.GetVar(ForumPath, "INFO", "CantMSG", ref argEmptySpaces)));

                // Elimino los post viejos
                var loopTo = NumPost;
                for (PostIndex = 1; PostIndex <= loopTo; PostIndex++)
                    FileSystem.Kill(
                        AppDomain.CurrentDomain.BaseDirectory + "Foros/" + withBlock.ID + PostIndex + ".for");


                var argEmptySpaces1 = 1024;
                NumPost = Convert.ToInt16(Migration.ParseVal(ES.GetVar(ForumPath, "INFO", "CantAnuncios",
                    ref argEmptySpaces1)));

                // Elimino los post viejos
                var loopTo1 = NumPost;
                for (PostIndex = 1; PostIndex <= loopTo1; PostIndex++)
                    FileSystem.Kill(AppDomain.CurrentDomain.BaseDirectory + "Foros/" + withBlock.ID + PostIndex +
                                    "a.for");


                // Elimino el foro
                FileSystem.Kill(AppDomain.CurrentDomain.BaseDirectory + "Foros/" + withBlock.ID + ".for");
            }
        }
    }

    public static bool SendPosts(short UserIndex, ref string ForoID)
    {
        bool SendPostsRet = default;
        // ***************************************************
        // Author: ZaMa
        // Last Modification: 22/02/2010
        // Sends all the posts of a required forum
        // ***************************************************

        short ForumIndex;
        short PostIndex;
        bool bEsGm;

        ForumIndex = GetForumIndex(ref ForoID);

        if (ForumIndex > 0)
        {
            {
                ref var withBlock = ref Foros[ForumIndex];

                // Send General posts
                var loopTo = (short)withBlock.CantPosts;
                for (PostIndex = 1; PostIndex <= loopTo; PostIndex++)
                {
                    ref var withBlock1 = ref withBlock.vsPost[PostIndex];
                    Protocol.WriteAddForumMsg(UserIndex,
                        (Declaraciones.eForumType)Declaraciones.eForumMsgType.ieGeneral, withBlock1.sTitulo,
                        withBlock1.Autor, withBlock1.sPost);
                }

                // Send Sticky posts
                var loopTo1 = (short)withBlock.CantAnuncios;
                for (PostIndex = 1; PostIndex <= loopTo1; PostIndex++)
                {
                    ref var withBlock2 = ref withBlock.vsAnuncio[PostIndex];
                    Protocol.WriteAddForumMsg(UserIndex,
                        (Declaraciones.eForumType)Declaraciones.eForumMsgType.ieGENERAL_STICKY, withBlock2.sTitulo,
                        withBlock2.Autor, withBlock2.sPost);
                }
            }

            bEsGm = Extra.EsGM(UserIndex);

            // Caos?
            if (Extra.esCaos(UserIndex) | bEsGm)
            {
                var argsForoID = FORO_CAOS_ID;
                ForumIndex = GetForumIndex(ref argsForoID);

                {
                    ref var withBlock3 = ref Foros[ForumIndex];

                    // Send General Caos posts
                    var loopTo2 = (short)withBlock3.CantPosts;
                    for (PostIndex = 1; PostIndex <= loopTo2; PostIndex++)
                    {
                        ref var withBlock4 = ref withBlock3.vsPost[PostIndex];
                        Protocol.WriteAddForumMsg(UserIndex,
                            (Declaraciones.eForumType)Declaraciones.eForumMsgType.ieCAOS, withBlock4.sTitulo,
                            withBlock4.Autor, withBlock4.sPost);
                    }

                    // Send Sticky posts
                    var loopTo3 = (short)withBlock3.CantAnuncios;
                    for (PostIndex = 1; PostIndex <= loopTo3; PostIndex++)
                    {
                        ref var withBlock5 = ref withBlock3.vsAnuncio[PostIndex];
                        Protocol.WriteAddForumMsg(UserIndex,
                            (Declaraciones.eForumType)Declaraciones.eForumMsgType.ieCAOS_STICKY, withBlock5.sTitulo,
                            withBlock5.Autor, withBlock5.sPost);
                    }
                }
            }

            // Caos?
            if (Extra.esArmada(UserIndex) | bEsGm)
            {
                var argsForoID1 = FORO_REAL_ID;
                ForumIndex = GetForumIndex(ref argsForoID1);

                {
                    ref var withBlock6 = ref Foros[ForumIndex];

                    // Send General Real posts
                    var loopTo4 = (short)withBlock6.CantPosts;
                    for (PostIndex = 1; PostIndex <= loopTo4; PostIndex++)
                    {
                        ref var withBlock7 = ref withBlock6.vsPost[PostIndex];
                        Protocol.WriteAddForumMsg(UserIndex,
                            (Declaraciones.eForumType)Declaraciones.eForumMsgType.ieREAL, withBlock7.sTitulo,
                            withBlock7.Autor, withBlock7.sPost);
                    }

                    // Send Sticky posts
                    var loopTo5 = (short)withBlock6.CantAnuncios;
                    for (PostIndex = 1; PostIndex <= loopTo5; PostIndex++)
                    {
                        ref var withBlock8 = ref withBlock6.vsAnuncio[PostIndex];
                        Protocol.WriteAddForumMsg(UserIndex,
                            (Declaraciones.eForumType)Declaraciones.eForumMsgType.ieREAL_STICKY, withBlock8.sTitulo,
                            withBlock8.Autor, withBlock8.sPost);
                    }
                }
            }

            SendPostsRet = true;
        }

        return SendPostsRet;
    }

    public static bool EsAnuncio(byte ForumType)
    {
        bool EsAnuncioRet = default;
        // ***************************************************
        // Author: ZaMa
        // Last Modification: 22/02/2010
        // Returns true if the post is sticky.
        // ***************************************************
        switch (ForumType)
        {
            case (byte)Declaraciones.eForumMsgType.ieCAOS_STICKY:
            {
                EsAnuncioRet = true;
                break;
            }

            case (byte)Declaraciones.eForumMsgType.ieGENERAL_STICKY:
            {
                EsAnuncioRet = true;
                break;
            }

            case (byte)Declaraciones.eForumMsgType.ieREAL_STICKY:
            {
                EsAnuncioRet = true;
                break;
            }
        }

        return EsAnuncioRet;
    }

    public static byte ForumAlignment(byte yForumType)
    {
        byte ForumAlignmentRet = default;
        // ***************************************************
        // Author: ZaMa
        // Last Modification: 01/03/2010
        // Returns the forum alignment.
        // ***************************************************
        switch (yForumType)
        {
            case (byte)Declaraciones.eForumMsgType.ieCAOS:
            case (byte)Declaraciones.eForumMsgType.ieCAOS_STICKY:
            {
                ForumAlignmentRet = Convert.ToByte((byte)Declaraciones.eForumType.ieCAOS);
                break;
            }

            case (byte)Declaraciones.eForumMsgType.ieGeneral:
            case (byte)Declaraciones.eForumMsgType.ieGENERAL_STICKY:
            {
                ForumAlignmentRet = Convert.ToByte((byte)Declaraciones.eForumType.ieGeneral);
                break;
            }

            case (byte)Declaraciones.eForumMsgType.ieREAL:
            case (byte)Declaraciones.eForumMsgType.ieREAL_STICKY:
            {
                ForumAlignmentRet = Convert.ToByte((byte)Declaraciones.eForumType.ieREAL);
                break;
            }
        }

        return ForumAlignmentRet;
    }

    public static void ResetForums()
    {
        // ***************************************************
        // Author: ZaMa
        // Last Modification: 22/02/2010
        // Resets forum info
        // ***************************************************
        // UPGRADE_WARNING: Es posible que la matriz Foros necesite tener elementos individuales inicializados. Haga clic aquí para obtener más información: 'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="B97B714D-9338-48AC-B03F-345B617E2B02"'
        // UPGRADE_WARNING: El límite inferior de la matriz Foros ha cambiado de 1 a 0. Haga clic aquí para obtener más información: 'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="0F1C9BE1-AF9D-476E-83B1-17D43BECFF20"'
        Foros = new tForo[2];
        ArrayInitializers.InitializeStruct(Foros);
        NumForos = 0;
    }

    private static void MoveArray(short ForumIndex, bool Sticky)
    {
        int i;

        {
            ref var withBlock = ref Foros[ForumIndex];
            if (Sticky)
                for (i = withBlock.CantAnuncios; i >= 2; i -= 1)
                {
                    withBlock.vsAnuncio[i].sTitulo = withBlock.vsAnuncio[i - 1].sTitulo;
                    withBlock.vsAnuncio[i].sPost = withBlock.vsAnuncio[i - 1].sPost;
                    withBlock.vsAnuncio[i].Autor = withBlock.vsAnuncio[i - 1].Autor;
                }
            else
                for (i = withBlock.CantPosts; i >= 2; i -= 1)
                {
                    withBlock.vsPost[i].sTitulo = withBlock.vsPost[i - 1].sTitulo;
                    withBlock.vsPost[i].sPost = withBlock.vsPost[i - 1].sPost;
                    withBlock.vsPost[i].Autor = withBlock.vsPost[i - 1].Autor;
                }
        }
    }

    public struct tPost
    {
        public string sTitulo;
        public string sPost;
        public string Autor;
    }

    public struct tForo
    {
        [VBFixedArray(MAX_MENSAJES_FORO)] public tPost[] vsPost;
        [VBFixedArray(MAX_ANUNCIOS_FORO)] public tPost[] vsAnuncio;
        public byte CantPosts;
        public byte CantAnuncios;
        public string ID;

        // UPGRADE_TODO: Se debe llamar a "Initialize" para inicializar instancias de esta estructura. Haga clic aquí para obtener más información: 'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="B4BFF9E0-8631-45CF-910E-62AB3970F27B"'
        public void Initialize()
        {
            // UPGRADE_WARNING: El límite inferior de la matriz vsPost ha cambiado de 1 a 0. Haga clic aquí para obtener más información: 'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="0F1C9BE1-AF9D-476E-83B1-17D43BECFF20"'
            vsPost = new tPost[MAX_MENSAJES_FORO + 1];
            // UPGRADE_WARNING: El límite inferior de la matriz vsAnuncio ha cambiado de 1 a 0. Haga clic aquí para obtener más información: 'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="0F1C9BE1-AF9D-476E-83B1-17D43BECFF20"'
            vsAnuncio = new tPost[MAX_ANUNCIOS_FORO + 1];
        }
    }
}